ocaml/testsuite/tests/typing-modes/legacy.ml

```diff
@@ -0,0 +1,116 @@
+(* TEST
```

Comment:
Maybe it belongs here, maybe in a different one: should there be a test that tries to create a new constructor of an extensible type within a portable function? I think that probably should not be allowed, because adding a new constructor of an extensible type causes a mutation of a hidden global variable (and could potentially cause a race).

Reply:
Usually such things are modelled by the function closing over a `uncontended` variable. But here, the variable being mutated is implicit, so there would require speical treatment.

Any chance that this global state is already made multithread-safe by ocaml runtime? That's likely considering that exception is a built-in feature and ocaml5 should take care of that.

Reply:
The underlying global variable is accessed using atomic primitives in OCaml 5.x. So I think it does not count as a race.

Reply:
I’d love @lpw25 ’s opinion here, as I don’t know that we have a written-down definition of “race”. (Maybe we do. I just don’t know where to find it.) That is, the extant primitives may prevent direct corruption, but there is still the possibility of unintended inter-thread communication via this channel. Is it possible that some function adds 10 constructors to an extensible type and then assumes that the tags are sequential? Probably not, but it’s conceivable to me.

Basically, my question boils down to: does “race-free” mean that the only inter-thread communication channels are our primitives? or does it only mean that there will be no data corruption as might be caused by tearing non-atomic structures?

Reply:
This is out of the scope of this PR (I want to merge this PR ASAP to avoid blocking others). Let's move this discussion to other venues.